hide invisible fields from email output according to visibility rules, fixes #1131

xini opened the pull request.

xini committed:
ENH Add config to hide invisible fields from email output

michalkleiner commented on code/Model/Submission/SubmittedFormField.php:
```diff
@@ -19,7 +19,8 @@ class SubmittedFormField extends DataObject
         'Value' => 'Text',
-        'Title' => 'Varchar(255)'
+        'Title' => 'Varchar(255)',
+        'IsVisible' => 'Boolean',
```
This is not needed, right? The field will carry it from the controller without being backed up by a true db field.

xini replied:
That's correct. I was on the fence here as well. But in the end I thought it might be beneficial to have that information available later on, maybe for an data export or similar?

michalkleiner replied:
Since it can be always calculated from the data and there's no visible UI to manage it, I'm for not having it.

xini replied:
hm.. the visibility could change over time, plus you would have to re-create the submitted form `$data` array to use `EditableFormField::isDisplayed()` to recalculate it. I think storing the flag is the better option? But if this is what it takes to get this accepted, I'll remove it.

ChloeSartorelli replied:
If the visibility state of the form changes after the form was submitted, the newly calculated value will not be correct for the submission. I'm in favor of it being stored too.

michalkleiner replied:
Ok, fair enough. Can we consider naming it `Displayed` then since it reflects a past state?

michalkleiner commented:
A sentence of text for the docs/changelog would be also good since some users may rely on all fields being in the export/email regardless of their display rules, so they would need to provide a custom template after this change.

GuySartorelli commented:
> Some users may rely on all fields being in the export/email regardless of their display rules, so they would need to provide a custom template after this change.

That's a good point. It wouldn't hurt to add a configuration option (perhaps on the `EmailRecipient` next to `HideFormData`) to allow showing all fields regardless of form visibility rules (or perhaps it should show all by default as that's the current behaviour, and then this option would allow respecting form visibility rules in the email).

GuySartorelli commented:
Also, we should probably mention in a comment on this PR that this fixes #1131 so there is a link between this and the issue.

xini commented:
Thanks @GuySartorelli and @michalkleiner for your feedback.

I have
- renamed the field on `SubmittedFormField` to `Displayed`
- reversed the changes to the templates to keep the current behaviour per default
- added a config flag `HideInvisibleFields` to `EmailRecipient` to only include visible fields in the generated email. Per default this flag is `false` making sure the current behaviour is maintained for existing users

These changes make sure that the current behaviour is maintained unless someone changes the settings on the recipient.

Where in the docs do you want me to mention this? There is no changelog file per se.

GuySartorelli commented:
That's great work! Thank you.

> Where in the docs do you want me to mention this?

I'd probably recommend somewhere in https://github.com/silverstripe/silverstripe-userforms/blob/5/docs/en/userguide/form-submissions.md#email-details

xini commented:
Would it make sense to set the default of `HideInvisibleFields` on `EmailRecipient` to `true`, so that newly created recipients get the new behaviour?

xini commented:
> I'd probably recommend somewhere in https://github.com/silverstripe/silverstripe-userforms/blob/5/docs/en/userguide/form-submissions.md#email-details

done.

GuySartorelli commented:
> Would it make sense to set the default of `HideInvisibleFields` on `EmailRecipient` to `true`, so that newly created recipients get the new behaviour?

I don't think so. It would be a departure from behaviour users are currently expecting and may even rely on. I can imagine cases for example where there are default values (perhaps hidden because the user filling in the form didn't tick a box that lets them edit the values) in a submission that need to be passed on to someone in the business who wasn't directly involved with setting up the form.

That said I'd defer to @michalkleiner in this case.

xini commented:
@GuySartorelli That's fine with me, no problem. Anything else to get this merged?

GuySartorelli commented:
That will be up to @michalkleiner or one of the other core committers - looks like the tests need fixing though.

michalkleiner commented on code/Control/UserDefinedFormController.php:
```diff
@@ -236,6 +236,7 @@ public function process($data, $form)
 
         $attachments = [];
         $submittedFields = ArrayList::create();
+        $visibleSubmittedFields = ArrayList::create();
```
```suggestion
```

michalkleiner commented on code/Control/UserDefinedFormController.php:
```diff
@@ -303,6 +307,10 @@ public function process($data, $form)
             }
 
             $submittedFields->push($submittedField);
+
+            if ($submittedField->Displayed) {
+                $visibleSubmittedFields->push($submittedField);
+            }
```
```suggestion
```

michalkleiner commented on code/Control/UserDefinedFormController.php:
```diff
@@ -350,6 +358,10 @@ public function process($data, $form)
                 $emailData['Body'] = SSViewer::execute_string($recipient->getEmailBodyContent(), $mergeFields);
+                // only include visible fields if recipient visibility flag is set
+                if ((bool) $recipient->HideInvisibleFields) {
+                    $emailData['Fields'] = $visibleSubmittedFields;
```
```suggestion
                    $emailData['Fields'] = $emailData['Fields']->filter('Displayed', true);
```

xini replied:
This only works for the first recipient. After that `$emailData['Fields']` is changed and can't be set back to all fields. That's why we need two arrays to store both versions of fields.

michalkleiner replied:
Ok, fair enough. Basically I want to avoid pushing each item separately into a new array, but we can use a variable and assign the filtered list to it for reuse.

xini replied:
Fair enough. That's changed now.

xini commented:
The Behat tests fail here: https://github.com/silverstripe/silverstripe-userforms/blob/5/tests/behat/features/userforms.feature#L144

Any ideas why? I don't get it.

xini commented:
> The Behat tests fail here: https://github.com/silverstripe/silverstripe-userforms/blob/5/tests/behat/features/userforms.feature#L144
> 
> Any ideas why? I don't get it.

The failing test doesn't seem to be caused by my changes, https://github.com/silverstripe/silverstripe-userforms/pull/1101/checks?check_run_id=5492750488 is failing too. 

Can this be merged and tagged please?

michalkleiner commented:
Yep, it seems the failure is not related to these changes.
@xini, can you please squash the commits into one with the commit message of "ENH Add config to hide invisible fields from email output". I'll merge it afterwards. Thanks!

xini commented:
@michalkleiner done, thanks!

xini commented:
@michalkleiner great, thanks so much!
